osrf/mbzirc
# Enable UAV flight with gripper

depends on #89

Main changes:
* Copied MulticopterControl system to mbzirc repo. I had to fix its computation of inertial values to account for a nested model (gripper). [These are the only new changes](https://github.com/osrf/mbzirc/blob/9f1b5c672d769c33d5418c1e24ecb8c195fa3b6a/mbzirc_ign/src/multicopter_control/MulticopterVelocityControl.cc#L310-L335). Changes will be upstreamed later and this custom system will be removed.
* Tuned motor params - this only works properly when vehicle inertial values are correctly computed with nested models.
* Fixed inertial values for sensor models
* Increased suction gripper weight (to be closer to finger gripper)


I added a world with a rig for testing UAV flight with 1kg payload. Make sure to **clear your fuel cache** as item mass has changed.

To test:

1. lauch sim **paused**

    ```
    ros2 launch ros_ign_gazebo ign_gazebo.launch.py ign_args:="-v 4 test/simple_demo_manip.sdf"
    ```

1. spawn quad with finger gripper

    ```
    ros2 launch mbzirc_ign spawn.launch.py name:=quadrotor_1 world:=simple_demo model:=mbzirc_quadrotor type:=uav x:=-0  y:=2.05 z:=1.92 R:=0 P:=0 Y:=0 slot3:=mbzirc_hd_camera gripper:=mbzirc_oberon7_gripper
    ```

1. Hit Play. A small 1kg black cube should fall on top  of the UAV.

1. Now try to make it take off with some linear z vel

    ```
    ros2 topic pub --once /quadrotor_1/cmd_vel geometry_msgs/msg/Twist '{linear: {x: 0.0, y: 0.0, z: 5}, angular: {x: 0.0, y: 0.0, z: 0.0}}'
    ```

You can also repeat the steps with a suction gripper and a target object. 

1. Spawn quad with suction gripper
   
     ```
    ros2 launch mbzirc_ign spawn.launch.py name:=quadrotor_1 world:=simple_demo model:=mbzirc_quadrotor type:=uav x:=-0  y:=2.05 z:=1.92 R:=0 P:=0 Y:=0 slot3:=mbzirc_hd_camera gripper:=mbzirc_suction_gripper
    ```

1. Hit Play. A small 1kg black cube should fall on top of the UAV. Use transform control tool to move the cube away

1. Enable suction to grab the small blue item

    ```
    ros2 topic pub --once /quadrotor_1/gripper/suction_on std_msgs/msg/Bool 'data: True'
    ```

1. Now take off!
    
    ```
    ros2 topic pub --once /quadrotor_1/cmd_vel geometry_msgs/msg/Twist '{linear: {x: 0.0, y: 0.0, z: 5}, angular: {x: 0.0, y: 0.0, z: 0.0}}'
    ```


![uav_suction_gripper_small](https://user-images.githubusercontent.com/4000684/163530710-da7a5c3f-9777-47f8-8240-ccd3d7b50b54.gif)
